plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -631,6 +636,32 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
                 .build();
     }
 
+    @Override
+    public void validateScan(ConnectorSession session, ConnectorTableHandle handle)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) handle;
+        Optional<PartitionSpec> partitionSpec = table.getPartitionSpecJson()
+                .map(partitionSpecJson -> PartitionSpecParser.fromJson(SchemaParser.fromJson(table.getTableSchemaJson()), partitionSpecJson));
+        if (partitionSpec.isEmpty() || partitionSpec.get().isUnpartitioned()) {
+            return;
+        }
+        if (isQueryPartitionFilterRequired(session)
+                && table.getEnforcedPredicate().isAll()) {
+            Set<Integer> columnsWithPredicates = table.getConstraintColumns().stream()
+                    .map(IcebergColumnHandle::getId)
+                    .collect(toImmutableSet());
+            Set<Integer> partitionColumns = partitionSpec.get().fields().stream()
+                    .filter(field -> !field.transform().isVoid())
+                    .map(PartitionField::sourceId)
+                    .collect(toImmutableSet());
+            if (Collections.disjoint(columnsWithPredicates, partitionColumns)) {
+                throw new TrinoException(
+                        StandardErrorCode.QUERY_REJECTED,
+                        format("Filter required on %s for at least one partition column: %s", table, partitionColumns));
```

Comment:
`partitionColumns` is an integer list. I don't think it's helpful for users. We should use column names in the error message.